src/utilities/new-sequence-editor/grammar.test.ts

```diff
@@ -1,64 +1,456 @@
+Command(Stem,⚠),
+Command(Stem,Args)
+))`,
+  ],
+  [
+    'Stem ending in disallowed character',
+    `FSW_CMD%`,
+    `Sequence(Commands(
+Command(Stem,⚠,Args)
+))`,
+  ],
+  [
+    'Mismatched brackets',
+    `CMD [[]
+CMD2 [
+CMD3 ]`,
+    `Sequence(Commands(
+Command(Stem,Args(RepeatArg(⚠))),
+Command(Stem,Args(RepeatArg(⚠,Enum)))
+))`,
+  ],
+  ['locals with wrong value types', `@LOCALS "string_not_enum"`, `Sequence(LocalDeclaration(⚠(String)))`],
+];
+
+const timeFormatTests = [
+  [
+    'Different absolute time tags',
+    `C CMD_1
+  A2030-001T12:34:56 CMD_2 "hello" "world"
+  A2030-001T12:34:56.789 CMD_3 "sub seconds"
+  `,
+    `Sequence(Commands(
+Command(
+TimeTag(TimeComplete),
+Stem,
+Args
+),
+Command(
+TimeTag(TimeAbsolute),
+Stem,
+Args(String,String)
+),
+Command(
+TimeTag(TimeAbsolute),
+Stem,
+Args(String)
+),
+))`,
+  ],
+  [
+    'Different relative times',
+    `R010T01:00:00.000 CMD_4 "hello" 10
+R00:10:00.000 CMD_5
+R00:00:01 CMD_6
+R00:00:01.123 CMD_7
+R10 CMD_8 10
+R10.123 CMD_9 10`,
+    `Sequence(Commands(
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args(String,Number)
+),
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args
+),
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args
+),
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args
+),
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args(Number)
+),
+Command(
+TimeTag(TimeRelative),
+Stem,
+Args(Number)
+)
+))`,
+  ],
+  [
+    'Different epoch times',
+    `E123T12:34:56.789 CMD_3
+E-123T12:34:56.789 CMD_3
+E+123T12:34:56.789 CMD_3
+E12:34:56.789 CMD_3
+E-12:34:56.789 CMD_3
+E+12:34:56.789 CMD_3
+E123T12:34:56 CMD_3
+E-123T12:34:56 CMD_3
+E+123T12:34:56 CMD_3
+E12:34:56 CMD_3
+E-12:34:56 CMD_3
+E+12:34:56 CMD_3
+E123 CMD_3
+E-123 CMD_3
+E+123 CMD_3
+E123.456 CMD_3
+E-123.456 CMD_3
+E+123.456 CMD_3
+  `,
+    `Sequence(Commands(
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args),
+Command(TimeTag(TimeEpoch),Stem,Args)
+  ))`,
+  ],
+];
+
+describe.each([
+  ['error tokens', errorTests],
+  ['parse tree structure', parseTreeTests],
+  ['time formats', timeFormatTests],
+])('grammar tests - %s', (_name: string, testArray: string[][]) => {
+  test.each(testArray)('%s', (_: string, input: string, expected: string) => {
+    // testTree will throw if there's a mismatch
+    expect(testTree(SeqLanguage.parser.parse(input), expected, undefined)).toBeUndefined();
```

Comment:
If it's expected to throw an error if it doesn't match, would it be more appropriate to use https://vitest.dev/api/expect.html#tothrowerror?

Also would you mind adding test cases for when parsing succeeds as well as errors?

Reply:
All of these tests will succeed, at least in that the grammar always produces a parse tree.

When the sample input doesn't conform to the language definition error tokens ("⚠") are emitted and included in the compared output

e.g. https://github.com/NASA-AMMOS/aerie-ui/blob/d20c44cfb5ec4c3568b1fd2b9428a27f0afb399b/src/utilities/new-sequence-editor/grammar.test.ts#L306

Reply:
I meant more along the lines of including explicit failure cases to actually test that an error is thrown. 

I think I see where my confusion is. Does `testTree` have to be used in order to actually test it? I got confused reading the test because I wasn't sure what the test was testing when expected the output to be `undefined`. Is this rewritable to something along the lines of:
```expect(SeqLanguage.parser.parse(input).toEqual(expected)```